src/hpa.c

```diff
@@ -538,6 +538,16 @@ hpa_shard_maybe_do_deferred_work(tsdn_t *tsdn, hpa_shard_t *shard,
 			purged = hpa_try_purge(tsdn, shard);
 			if (purged) {
 				nops++;
+			} else {
+				/*
+				 * It's important to explicitly break on
+				 * failure, as otherwise it's possible to end
+				 * up in an infinite loop continuously trying
+				 * to purge without actually being able to
+				 * do so.
+				 * See GitHub issue #2449.
```

Comment:
nit: Let's avoid referring to github issues in the source / comments. Put in commit message like you have right now is sufficient.

Reply:
Actually come to think of it, I don't know that this is the right fix. I just pulled this up from a conversation on this topic with @davidtgoldblatt when I had encountered a similar issue, and he said:

> I think that combination of hpa_hugification_threshold_ratio and hpa_dirty_mult doesn't make sense
the one says "create 25 percent more dirty pages if you have to in order to hugify", and the other says "purge if you're more than 3 percent dirty (say, because you just created a bunch of dirty pages by hugifying)"
(probably there should be a warning or something if their sum is < 1)

Reply:
I don't have a strong opinion. I guess stylistically, I'd say that it's hpa_should_purge that should be returning false in this situation, rather than having this loop decline to purge even when it's told it ought to.

Like, at its core this is a misconfiguration issue[1]. Our choices for responding to misconfiguration issues are:
1. Silently ignore the settings and pick a reasonable default as an alternative
2. Noisily ignore the settings and pick a reasonable default as an alternative
3. Crash at boot
4. Perform pathologically badly

I think all of these have their pros/cons. I'd maybe do 2 by default or 3 when abort_conf is true? Maybe what I'm saying is that this should be handled at conf parsing time (with a crash here after the fact if conf parsing didn't fix it). No strong opinions though, really.

[1] Not really blaming the users; it's defaults that are broken here.

Reply:
Implemented in the latest push. Let me know what you think, and once we're happy with everything I'll add test-cases exercising this new behavior.